src/ui/component/toast/toast.tsx

```diff
@@ -1,51 +1,140 @@
+import { Icon } from '@/ui/component/icon/icon'
+import { useOnClickOutside } from '@/ui/hook/useOnClickOutside'
+import { useOnKeyboard } from '@/ui/hook/useOnKeyboard'
+import { Button } from '@/ui/component//button/button'
+import type { DismissNotificationInput } from '@/domain/notification/aggregate'
+import '@/ui/component/notifications/i18n'
 import './toast.scss'
 
-type ToastVariants = 'success' | 'error' | 'warning' | 'info'
+type ToastCTAProps = {
+  type: NotificationType
+  action: string
+}
+const ToastCTA: FC<ToastCTAProps> = ({ action, type }) => {
+  const { t } = useTranslation('notifications')
+
+  const handleClick = useCallback(() => {
+    window.location.reload()
+  }, [])
+  return (
+    <Button
+      className={classNames('okp4-dataverse-portal-toast-cta', type)}
+      label={t(`action.${action}`)}
+      onClick={handleClick}
+    />
+  )
+}
 
 type ToastProps = {
+  id: NotificationID
+  type: NotificationType
   title: string
-  autoHideDuration?: number
-  description?: string
-  icon?: JSX.Element
-  onClose?: () => void
-  open?: boolean
-  variant?: ToastVariants
+  onClose: (input: DismissNotificationInput) => void
+  message?: string
+  iconName?: IconName
+  action?: string
+  index: number
 }
 
-export const Toast = ({
+// eslint-disable-next-line max-lines-per-function
+export const Toast: FC<ToastProps> = ({
+  id,
+  type,
   title,
-  autoHideDuration,
-  description,
-  icon,
   onClose,
-  open = false,
-  variant = 'info'
-}: ToastProps): JSX.Element | null => {
+  message,
+  iconName,
+  action,
+  index
+}) => {
+  const [isHidding, setIsHidding] = useState(false)
+  const containerRef = useRef<HTMLDivElement | null>(null)
+  const timeoutRef = useRef<ReturnType<typeof setTimeout>>()
+  const autoHideTimeoutRef = useRef<ReturnType<typeof setTimeout>>()
+  const animationDurationRef = useRef<number>(500)
+  const stackedPosition = useRef<number>(index + 1)
+
+  const autoHideDuration = useMemo((): number => {
+    const messagesLength = title.length + (message?.length ?? 0)
+    const minDuration = 2000
+    const maxDuration = 7000
+    const durationPerChar = 50
+    const duration = messagesLength * durationPerChar
+    return Math.min(Math.max(duration, minDuration), maxDuration)
+  }, [title, message])
+
+  const handleAction = useCallback(() => {
+    if (isHidding) return
+    setIsHidding(true)
+  }, [isHidding])
+
+  useEffect(() => {
+    if (action) return
+    autoHideTimeoutRef.current = setTimeout(() => {
+      handleAction()
+    }, autoHideDuration)
+    return () => clearTimeout(autoHideTimeoutRef.current)
+  }, [handleAction, action, autoHideDuration])
+
   useEffect(() => {
-    if (open && autoHideDuration) {
-      const autoHideDurationTimer = setTimeout(() => {
-        onClose?.()
-      }, autoHideDuration)
-      return () => {
-        clearTimeout(autoHideDurationTimer)
+    if (!isHidding) return
+    timeoutRef.current = setTimeout(() => {
```

Comment:
I'm not sure you need to memoize the `setTimeout` return's value in a `React.Ref`.

The first version using a `constant` looked good to me and seems more straightforward , what do you think ?

Reply:
We don't need to you're right. Thanks!